### PR TITLE
[Core] Migrate `HostSession` logging to C++ and refactor

### LINKED ISSUE #531: Add logging to C++ `HostSession`
## What

Complete C++ `HostSession` migration, remove Python override, and make `final`.

## ACs

- [x] `severity` and `message` arguments to `LoggerInterface::log` swapped (independent PR).
- [x] `HostSession` removed from Python, replaced by bound `final` C++ class.
- [x] Log abstraction removed, and `logger()` method added that returns the `LoggerInterface` instance.
- [x] Severity mirror removed.
- [x] `HostSession` made final.
- [x] Python `HostSession` implementation removed.

### PR DESCRIPTION
**Pull request by foundrytom**
Removes the lingering Python `HostSession` wrapper, taking the opportunity to refactor the approach as described in the issue.

Noticed an inadvertent `const` left on `LoggerInterface::log`, which is removed in the first commit.

Closes #531.

### CODE REVIEW

File: tests/python/conftest.py [hunk 1]
```diff
@@ -78,7 +78,7 @@ def mock_host_session(mock_host_interface, mock_logger):
     Fixture for a `HostSesssion`, configured with a mock_host_interface,
     that forwards method calls to an internal public `mock.Mock` instance.
     """
-    return MockHostSession(Host(mock_host_interface), mock_logger)
+    return HostSession(Host(mock_host_interface), mock_logger)
```
Comment (feltech) [inline tests/python/conftest.py:81]: [consider][minor] `test_hostsession.py` duplicates this fixture somewhat (though it splits out `a_host` to its own fixture)

File: src/openassetio-core/hostApi/ManagerFactory.cpp [hunk 2]
```diff
@@ -42,7 +42,7 @@ ManagerFactory::ManagerDetails ManagerFactory::availableManagers() const {
 
   // `HostSession` required for `settings()` calls.
   const managerApi::HostSessionPtr hostSession =
-      managerApi::HostSession::make(managerApi::Host::make(hostInterface_));
+      managerApi::HostSession::make(managerApi::Host::make(hostInterface_), logger_);
```
Comment (feltech) [inline src/openassetio-core/hostApi/ManagerFactory.cpp:45]: [fix] Looks like was an oversight during development of #507, where we initially had `settings` as part of the `ManagerDetail`, and so required a `HostSession`.  We removed that but missed removing this line.

**Review by feltech — CHANGES_REQUESTED** [review]
Looks good, just a couple of very minor notes.

File: tests/python/conftest.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (foundrytom) [reply tests/python/conftest.py]: Good call, renamed [here](https://github.com/OpenAssetIO/OpenAssetIO/pull/541/commits/96c8b96ff2efa408ca098c6abf4422b7ba0eac5c).

Comment (foundrytom) [reply tests/python/conftest.py]: Switched to using the shared fixture(s) [here](https://github.com/OpenAssetIO/OpenAssetIO/pull/541/commits/fc13ea7c654728678dd1c5ad522ffe65f8bfbf67).

File: src/openassetio-core/hostApi/ManagerFactory.cpp (hunk 2 above)
Comment (foundrytom) [reply src/openassetio-core/hostApi/ManagerFactory.cpp]: Good spot, removed [here](https://github.com/OpenAssetIO/OpenAssetIO/pull/541/commits/03ce704860e1ac773242977a708bfc1b86a0b72f). Curious none of our linters pick up 'unused variable'.

File: tests/python/conftest.py (hunk 1 above)
Comment (feltech) [reply tests/python/conftest.py]: Looks good, just the pylint failure due to this removal (noticed via the github annotations :smile: )

File: src/openassetio-core/hostApi/ManagerFactory.cpp (hunk 2 above)
Comment (feltech) [reply src/openassetio-core/hostApi/ManagerFactory.cpp]: Good point.  I think the `make` functions should be marked `[[nodiscard]]`, I bet the linter(s) would complain then.

File: tests/python/conftest.py (hunk 1 above)
Comment (feltech) [reply tests/python/conftest.py]: ^ ah OK, fixed in [75dc42b](https://github.com/OpenAssetIO/OpenAssetIO/pull/541/commits/75dc42b18c711a7e36324acfff0156f0ae65c6d4)

File: src/openassetio-core/hostApi/ManagerFactory.cpp (hunk 2 above)
Comment (foundrytom) [reply src/openassetio-core/hostApi/ManagerFactory.cpp]: Ahh yes, sorry I forgot that when we did it.

**Review by feltech — APPROVED** [review]
LGTM squash and merge when ready

File: src/openassetio-core/hostApi/ManagerFactory.cpp (hunk 2 above)
Comment (foundrytom) [reply src/openassetio-core/hostApi/ManagerFactory.cpp]: @feltech  added https://github.com/OpenAssetIO/OpenAssetIO/issues/544

**Pull request merged**